client/core/core_test.go

```diff
@@ -2624,10 +2635,142 @@ func TestHandlePreimageRequest(t *testing.T) {
 			)
 		}
 		tracker.mtx.RUnlock()
+	})
+	t.Run("more than one preimage request for order (different csums)", func(t *testing.T) {
+		rig := newTestRig()
+		defer rig.shutdown()
+		ord := &order.LimitOrder{P: order.Prefix{ServerTime: time.Now()}}
+		oid := ord.ID()
+		preImg := newPreimage()
+		firstCSum := dex.Bytes{2, 3, 5, 7, 11, 13}
+
+		tracker := &trackedTrade{
+			Order:  ord,
+			preImg: preImg,
+			mktID:  tDcrBtcMktName,
+			db:     rig.db,
+			dc:     rig.dc,
+			// Simulate first preimage request by initializing csum here.
+			csum:     firstCSum,
+			metaData: &db.OrderMetaData{},
+		}
+
+		// Test the new path with rig.core.sentCommits.
+		readyCommitment := func(commit order.Commitment) chan struct{} {
+			commitSig := make(chan struct{}) // close after fake order submission is "done"
+			rig.core.sentCommitsMtx.Lock()
+			rig.core.sentCommits[commit] = commitSig
+			rig.core.sentCommitsMtx.Unlock()
+			return commitSig
+		}
+
+		commit := preImg.Commit()
+		commitSig := readyCommitment(commit)
+		secondCSum := dex.Bytes{2, 3, 5, 7, 11, 14}
+		payload := &msgjson.PreimageRequest{
+			OrderID:        oid[:],
+			Commitment:     commit[:],
+			CommitChecksum: secondCSum,
+		}
+		reqCommit, _ := msgjson.NewRequest(rig.dc.NextID(), msgjson.PreimageRoute, payload)
+
+		notes := rig.core.NotificationFeed()
 
-		// TODO
-		// We shouldn't allow for a duplicate request to contain different csum (test against that),
-		// see https://github.com/decred/dcrdex/pull/1077#discussion_r641593845 for details.
+		rig.dc.trades[oid] = tracker
+		err := handlePreimageRequest(rig.core, rig.dc, reqCommit)
+		if err != nil {
+			t.Fatalf("handlePreimageRequest error: %v", err)
+		}
+
+		// It has gone async now, waiting for commitSig.
+		// i.e. "Received preimage request for %v with no corresponding order submission response! Waiting..."
+		close(commitSig) // pretend like the order submission just finished
+
+		select {
+		case note := <-notes:
+			t.Fatalf("got unexpected note: %v", note)
+		case <-time.After(time.Millisecond):
+			// We've waited enough for a note to come, hopefully it won't after
+			// this point in time.
+		}
```

Comment:
Reading through this code once again I've realized the following tests:
- `more than one preimage request for order (different csums)`
- `more than one preimage request for cancel order (different csums)`

are flaky as is currently implemented, 
I found some usage of `time.Sleep` in the tests in `dcrdex`, so I asume it is an accepted approach - "to allow for flakiness when it is hard/time-consuming to implement otherwise" ?

Should I try to address this, or we better move on ?

Reply:
Indeed sleeps in test are generally to be avoided, and were often a thorn in my side early on.  These "different csums" tests are tricky of course because you are verifying the _absence_ of an event, and the tested code is launched by `handlePreimageRequest` in a (async) goroutine.

You could hack a channel send into `(conn *TWebsocket) Send` perhaps, and receive on that like the other tests receive on the `notes` chan.  A 250ms wait here in this select isn't terrible either, but 1 ms is too short.  My only hesitation with the channel hacks, despite being an effective sync mechanism, is that they make the tests difficult to change in the future.

Another suggestion I have is to keep the NewResponse/Send in processPreimage, which will both reduce the input args to `acceptCsum` and allow a defer to unlock the tracker mutex.

The gist of the ideas:

```diff
diff --git a/client/core/core.go b/client/core/core.go
index e73c82b7..b2f9ce9b 100644
--- a/client/core/core.go
+++ b/client/core/core.go
@@ -5080,11 +5080,6 @@ func processPreimageRequest(c *Core, dc *dexConnection, reqID uint64, oid order.
                return fmt.Errorf("no active order found for preimage request for %s", oid)
        }
 
-       err := acceptCsum(dc, reqID, tracker, isCancel, commitChecksum)
-       if err != nil {
-               return fmt.Errorf("accept csum: %w", err)
-       }
-
        // Clean up the sentCommits now that we loaded the commitment. This can be
        // removed when the old piSyncers method is removed.
        defer func() {
@@ -5093,6 +5088,24 @@ func processPreimageRequest(c *Core, dc *dexConnection, reqID uint64, oid order.
                delete(c.sentCommits, preImg.Commit()) // redundant if the commitment was in request
                c.sentCommitsMtx.Unlock()
        }()
+
+       // Record the csum if this preimage request is novel, and deny it if this is
+       // a duplicate request with an altered csum.
+       if !acceptCsum(tracker, isCancel, commitChecksum) {
+               csumErr := errors.New("invalid csum in duplicate preimage request")
+               resp, err := msgjson.NewResponse(reqID, nil,
+                       msgjson.NewError(msgjson.InvalidRequestError, csumErr.Error()))
+               if err != nil {
+                       c.log.Errorf("Failed to encode response to denied preimage request: %v", err)
+                       return csumErr
+               }
+               err = dc.Send(resp)
+               if err != nil {
+                       c.log.Errorf("Failed to send response to denied preimage request: %v", err)
+               }
+               return csumErr
+       }
+
        resp, err := msgjson.NewResponse(reqID, &msgjson.PreimageResponse{
                Preimage: preImg[:],
        }, nil)
@@ -5114,48 +5127,27 @@ func processPreimageRequest(c *Core, dc *dexConnection, reqID uint64, oid order.
 // acceptCsum will record the commitment checksum so we can verify that the
 // subsequent match_proof with this order has the same checksum. If it does not,
 // the server may have used the knowledge of this preimage we are sending them
-// now to alter the epoch shuffle.
-func acceptCsum(dc *dexConnection, reqID uint64, tracker *trackedTrade, isCancel bool, commitChecksum dex.Bytes) error {
-       var respondWithErr bool
-
+// now to alter the epoch shuffle. The return value is false if a previous
+// checksum has been recorded that differs from the provided one.
+func acceptCsum(tracker *trackedTrade, isCancel bool, commitChecksum dex.Bytes) bool {
        // Do not allow csum to be changed once it has been committed to
        // (initialized to something other than `nil`) because it is probably a
        // malicious behavior by the server.
        tracker.mtx.Lock()
+       defer tracker.mtx.Unlock()
        if isCancel {
-               if tracker.cancel.csum != nil {
-                       if !bytes.Equal(commitChecksum, tracker.cancel.csum) {
-                               respondWithErr = true
-                       }
-               } else {
+               if tracker.cancel.csum == nil {
                        tracker.cancel.csum = commitChecksum
+                       return true
                }
-       } else {
-               if tracker.csum != nil {
-                       if !bytes.Equal(commitChecksum, tracker.csum) {
-                               respondWithErr = true
-                       }
-               } else {
-                       tracker.csum = commitChecksum
-               }
+               return bytes.Equal(commitChecksum, tracker.cancel.csum)
        }
-       tracker.mtx.Unlock()
 
-       if respondWithErr {
-               csumErr := errors.New("csum is already initialized")
-               resp, err := msgjson.NewResponse(reqID, nil,
-                       msgjson.NewError(msgjson.InvalidRequestError, csumErr.Error()))
-               if err != nil {
-                       return fmt.Errorf("preimage response encoding error: %w", err)
-               }
-               err = dc.Send(resp)
-               if err != nil {
-                       return fmt.Errorf("preimage send error: %w", err)
-               }
-               return csumErr
+       if tracker.csum == nil {
+               tracker.csum = commitChecksum
+               return true
        }
-
-       return nil
+       return bytes.Equal(commitChecksum, tracker.csum)
 }
```

```diff
diff --git a/client/core/core_test.go b/client/core/core_test.go
index 4eb1ae22..14982369 100644
--- a/client/core/core_test.go
+++ b/client/core/core_test.go
@@ -136,12 +136,13 @@ var (
 )
 
 type TWebsocket struct {
-       mtx        sync.RWMutex
-       id         uint64
-       sendErr    error
-       reqErr     error
-       connectErr error
-       msgs       <-chan *msgjson.Message
+       mtx            sync.RWMutex
+       id             uint64
+       sendErr        error
+       sendMsgErrChan chan *msgjson.Error
+       reqErr         error
+       connectErr     error
+       msgs           <-chan *msgjson.Message
        // handlers simulates a peer (server) response for request, and handles the
        // response with the msgFunc.
        handlers map[string][]func(*msgjson.Message, msgFunc) error
@@ -229,7 +230,20 @@ func (conn *TWebsocket) NextID() uint64 {
        conn.id++
        return conn.id
 }
-func (conn *TWebsocket) Send(msg *msgjson.Message) error { return conn.sendErr }
+func (conn *TWebsocket) Send(msg *msgjson.Message) error {
+       if conn.sendMsgErrChan != nil {
+               resp, err := msg.Response()
+               if err != nil {
+                       return err
+               }
+               if resp.Error != nil {
+                       conn.sendMsgErrChan <- resp.Error
+                       return nil // the response was sent successfully
+               }
+       }
+
+       return conn.sendErr
+}
 func (conn *TWebsocket) Request(msg *msgjson.Message, f msgFunc) error {
        return conn.RequestWithTimeout(msg, f, 0, func() {})
 }
@@ -2674,7 +2688,10 @@ func TestHandlePreimageRequest(t *testing.T) {
                }
                reqCommit, _ := msgjson.NewRequest(rig.dc.NextID(), msgjson.PreimageRoute, payload)
 
-               notes := rig.core.NotificationFeed()
+               // Prepare to have processPreimageRequest respond with a payload with
+               // the Error field set.
+               rig.ws.sendMsgErrChan = make(chan *msgjson.Error, 1)
+               defer func() { rig.ws.sendMsgErrChan = nil }()
 
                rig.dc.trades[oid] = tracker
                err := handlePreimageRequest(rig.core, rig.dc, reqCommit)
@@ -2687,11 +2704,12 @@ func TestHandlePreimageRequest(t *testing.T) {
                close(commitSig) // pretend like the order submission just finished
 
                select {
-               case note := <-notes:
-                       t.Fatalf("got unexpected note: %v", note)
-               case <-time.After(time.Millisecond):
-                       // We've waited enough for a note to come, hopefully it won't after
-                       // this point in time.
+               case msgErr := <-rig.ws.sendMsgErrChan:
+                       if msgErr.Code != msgjson.InvalidRequestError {
+                               t.Fatalf("expected error code %d got %d", msgjson.InvalidRequestError, msgErr.Code)
+                       }
+               case <-time.After(time.Second):
+                       t.Fatal("no msgjson.Error sent from preimage request handling")
                }
 
                tracker.mtx.RLock()
```

Reply:
> My only hesitation with the channel hacks, despite being an effective sync mechanism, is that they make the tests difficult to change in the future.

Noted on that. Since you've already sketched out the implementation, I'd rather use it here :)

> Another suggestion I have is to keep the NewResponse/Send in processPreimage, which will both reduce the input args to acceptCsum and allow a defer to unlock the tracker mutex.

Yeah, looks cleaner indeed, applied.